src/main/clojure/clara/rules/compiler.clj

```diff
@@ -1242,19 +1248,211 @@
+                            ;; anyway.
+                            (try
+                              (eval exprs)
+                              (catch Exception e
+                                ;; Using mapv here rather than map to avoid laziness, otherwise compilation failure might
+                                ;; fall into the throw below for the wrong reason.
+                                (mapv (fn [expr node-keys]
+                                        (let [cmeta (meta node-keys)]
+                                          (with-bindings
+                                            {#'*compile-ctx* (:compile-ctx cmeta)
+                                             #'*file* (:file cmeta *file*)}
+                                            (try-eval expr))))
+                                      exprs
+                                      node-keys)
+                                ;; If none of the rules are the issue, it is likely that the
+                                ;; size of the code trying to be evaluated has exceeded the limit
+                                ;; set by java.
+                                (throw (ex-info (str "There was a failure while batch evaling the node expressions, " \newline
+                                                     "but wasn't present when evaling them individually. This likely indicates " \newline
+                                                     "that the method size exceeded the maximum set by the jvm, see the cause for the actual error.")
+                                                {:node-keys node-keys}
+                                                e)))))]
+    (into {}
+          cat
+          ;; Grouping by ns, most expressions will not have a defined ns, only expressions from production nodes.
+          ;; These expressions must be evaluated in the ns where they were defined, as they will likely contain code
+          ;; that is namespace dependent.
+          (for [[nspace group] (sort-by key (group-by (comp :ns meta key) key->expr))
+                ;; Partitioning the number of forms to be evaluated, Java has a limit to the size of methods if we were
+                ;; evaluate all expressions at once it would likely exceed this limit and throw an exception.
+                expr-batch (partition-all partition-size group)
+                :let [node-keys (map key expr-batch)]]
+            (mapv vector
+                  node-keys
+                  (with-bindings (if nspace
+                                   {#'*ns* (the-ns nspace)}
+                                   {})
+                    (batching-try-eval node-keys (mapv val expr-batch))))))))
+
+(defn safe-get
+  "A helper function for retrieving a given key from the provided map. If the key doesn't exist within the map this
+   function will throw an exception."
+  [m k]
+  (let [v (get m k ::not-found)]
+    (if (identical? v ::not-found)
```

Comment:
I presume the reason you used identical? instead of = here was for performance?  @EthanEChristian : I realize that as we've discussed previously in another (non-Clara) context it is currently part of Clojure's implementation that equal keywords are identical, and so when comparing two things that are always keywords using identity checks is a way to gain marginally better performance.  However, the number of operations here is just the number of nodes in the rules session compiled, and then only at compile-time, correct?  I doubt that that will be any kind of serious problem for a number of checks (at compile time) on the order of thousands or tens of thousands, and this pattern isn't a part of Clara's codebase yet.  As is, this introduces a maintenance burden of keeping track of whether Clojure has changed this; I don't believe that property is part of any API guarantees.

In this particular case, I think we can resolve this one of two ways
- Using clojure.core/=.  I don't think this is likely to have any meaningful performance cost, but if you have data to the contrary please post.
- Creating a let-binding for the keyword so that it actually is guaranteed to be identical by Clojure's public contracts and commenting with the reasoning.  

i.e.

```
(defn safe-get
  "A helper function for retrieving a given key from the provided map. If the key doesn't exist within the map this
   function will throw an exception."
  [m k]
  (let [not-found-keyword ::not-found
        v (get m k not-found-keyword)]
    (if (identical? v not-found-keyword)
      (throw (ex-info "Key not found with safe-get" {:map map :key k}))
      v)))
```